### ISSUE TITLE
serialize_from_session on salt change

### ISSUE THREAD

**Issue by killthekitten** [plataformatec/devise#2527]
Currently, serialization from session depends on authenticatable salt, which is a part of `encrypted_password` field. And all goes smoothly while this field isn't changing (99% of cases). But when it does change, deserialization doesn't happen. As a result of this cases  technically you are being logged out:

```
def fetch(scope)
  key = session[key_for(scope)]
  return nil unless key

  method_name = "#{scope}_deserialize"
  user = respond_to?(method_name) ? send(method_name, key) : deserialize(key)
  delete(scope) unless user
  user
end

def serialize_from_session(key, salt)
  record = to_adapter.get(key)
  record if record && record.authenticatable_salt == salt
end
```

That's ok, but what is confusing me, is that every session variable except `warden.user.user.key` is still stored in the session. None of the hooks that clear session variables gets called.

I want to understand two things about that:
1. Does the session act the right way?
2. Is there any other real cases of authentication salt change except i.e. seed regeneration?

In other words, do we must fix this behavior of devise or we can keep it in it's current state?
